vm-validator/src/vm_validator.rs

```diff
@@ -121,3 +122,47 @@ pub fn get_account_sequence_number(
         None => Ok(0),
     }
 }
+
+// A pool of VMValidators that can be used to validate transactions concurrently. This is done because
+// the VM is not thread safe today. This is a temporary solution until the VM is made thread safe.
+#[derive(Clone)]
+pub struct PooledVMValidator {
+    vm_validators: Vec<Arc<Mutex<VMValidator>>>,
+}
+
+impl PooledVMValidator {
+    pub fn new(db_reader: Arc<dyn DbReader>, pool_size: usize) -> Self {
+        let mut vm_validators = Vec::new();
+        for _ in 0..pool_size {
+            vm_validators.push(Arc::new(Mutex::new(VMValidator::new(db_reader.clone()))));
+        }
+        PooledVMValidator { vm_validators }
+    }
+
+    pub fn get_next_vm(&self) -> Arc<Mutex<VMValidator>> {
+        let mut rng = thread_rng(); // Create a thread-local random number generator
+        let random_index = rng.gen_range(0, self.vm_validators.len()); // Generate random index
+        self.vm_validators[random_index].clone() // Return the VM at the random index
+    }
+}
+
+impl TransactionValidation for PooledVMValidator {
+    type ValidationInstance = AptosVM;
+
+    fn validate_transaction(&self, txn: SignedTransaction) -> Result<VMValidatorResult> {
+        self.get_next_vm().lock().unwrap().validate_transaction(txn)
+    }
+
+    fn restart(&mut self) -> Result<()> {
+        for vm_validator in &self.vm_validators {
+            vm_validator.lock().unwrap().restart()?;
+        }
```

Comment:
Not sure if this is blocking too long for waiting for all the locks. Can be made lazy by setting a version flag globally and inside each lock so the work is spread to multiple threads when they are gonna be used. Unnecessary if there's no performance concern of course.

Reply:
What you mean by "blocking too long for waiting for all locks"?  We are not blocking to acquire all the locks but attempting to lock them one by one.

Reply:
Yeah, I mean this function will wait for all locks sequentially. this function will be 32x slower than before if we have concurrency of 32.